### ISSUE TITLE
Concerns on the implementation of the relative positional embedding

### ISSUE THREAD

**Issue by PkuRainBow** [csrhddlam/axial-deeplab#1]
Really nice work! I have a small concern on the implementation details of the relative positional embedding.

This is the **overall random initialization** of the positional embedding:

https://github.com/csrhddlam/axial-deeplab/blob/17cdab8ecc785811821ffa57b89cbfe44b9b3063/lib/models/axialnet.py#L50-L53

This is the estimation of the relative positional embedding by taking **a slice of the overall random initialized vector**:
https://github.com/csrhddlam/axial-deeplab/blob/17cdab8ecc785811821ffa57b89cbfe44b9b3063/lib/models/axialnet.py#L76-L82

According to the above implementations, I guess you implement the **property of relative position** via sharing a fixed ratio (part) of the position embedding instead of subtracting one positional embedding from the other one.

**Besides, I am also concerned about the efficiency of the above loop. How could we improve the efficacy here?**

I am wondering why you choose this implementation and whether there exists any other work that has used similar implementations?

**Comment by phj128** [comment]
Thanks for your interest.

Here we do share a big relative position embedding and optimize this one. Your understanding is correct. BTW, what do you mean of subtracting one positional embedding from the other one? Like we have position embeddings for every pixel and substract them to get relative ones?

https://arxiv.org/abs/1906.05909 This paper uses a similar way but they did not release their code, either.

And your concerns are reasonable. However, I did not test the time cost here and I do not think it needs too much time. But your comments are right and this implementation is not efficient. I did this because it is easy to implement :) What's more, here we can actually init the embedding vectors only one time and optimize them, instead of "torch.cat" in every forward while if you implement in this way, there would be some other problems during backwards, like same relative position should share the gradients. I do not have a lot of experience here, so I just did in this way because it is easy :)

If you find any methods, please tell me. Also, I did not find other implementations about this part, but if you know any of those and share with me, I would appreciate it.
